File: materials/rust/orders/src/infrastructure.rs
use std::{
    fs::OpenOptions,
    io::Write,
    os::unix::fs::OpenOptionsExt,
    thread::sleep,
    time::Duration,
};

use crate::models::Order;

pub struct RandomSQLDatabase {
    pub connection_string: String,
}

impl RandomSQLDatabase {
    pub fn new() -> RandomSQLDatabase {
        RandomSQLDatabase {
            connection_string: "random://root:password@localhost:228/shop".to_string(),
        }
    }
    pub fn save_order(&self, order: &Order, total: f64) -> Result<(), std::io::Error> {
        println!("Connecting to RandomSQL at {}...", self.connection_string);
        sleep(Duration::from_millis(
            500 * 1, /*todo: time.Milliseconds */
        ));

        let mut file = OpenOptions::new()
            .append(true)
            .create(true)
            .write(true)
            .read(false)
            .mode(0o644)
            .open("orders_db.txt")?;

        // Can it be modelled in Rust?
        // defer file.Close()

        // TODO: a `chrono` or `time` crate is required to format time as in RFC3339
        let record = format!(
            "[{}] ID: {} | Type: {} | Total : {:.2}\n",
            "ONCE", order.id, order.r#type, total
        );
        file.write_all(record.as_bytes())?;
        println!("Order saved successfully.");
        Ok(())
    }
}

pub struct SmtpMailer {
    pub server: String,
}

impl SmtpMailer {
    pub fn send_html_email(&self, to: &str, subject: &str, body: &str) {
        println!(">> Connecting to SMTP server {}...", self.server);
        println!(">> Sending EMAIL to {to}\n   Subject: {subject}\n   Body: {body}");
    }
}

File: materials/orders/swift/Infrastructure.swift
import Foundation

final class RandomSQLDatabase {
    let connectionString: String

    init(connectionString: String) {
        self.connectionString = connectionString
    }

    static func newMySQLDatabase() -> RandomSQLDatabase {
        return RandomSQLDatabase(connectionString: "random://root:password@localhost:228/shop")
    }

    func saveOrder(_ order: Order, total: Double) throws {
        print("Connecting to RandomSQL at \(connectionString) ...")
        Thread.sleep(forTimeInterval: 0.5)

        let fileURL = URL(fileURLWithPath: "orders_db.txt")
        let record = "[\(rfc3339Now())] ID: \(order.id) | Type: \(order.type) | Total: \(String(format: "%.2f", total))\n"

        if FileManager.default.fileExists(atPath: fileURL.path) == false {
            FileManager.default.createFile(atPath: fileURL.path, contents: nil)
        }

        guard let data = record.data(using: .utf8) else { return }
        let handle = try FileHandle(forWritingTo: fileURL)
        defer { try? handle.close() }

        try handle.seekToEnd()
        try handle.write(contentsOf: data)

        print("Order saved successfully.")
    }

    private func rfc3339Now() -> String {
        let fmt = ISO8601DateFormatter()
        fmt.formatOptions = [.withInternetDateTime]
        return fmt.string(from: Date())
    }
}

final class SmtpMailer {
    let server: String

    init(server: String) {
        self.server = server
    }

    func sendHtmlEmail(to: String, subject: String, body: String) {
        print(">> Connecting to SMTP server \(server)...")
        print(">> Sending EMAIL to \(to)\n   Subject: \(subject)\n   Body: \(body)")
    }
}
